sdk/extensions/Microsoft.Extensions.Azure/src/Internal/ClientFactory.cs

```diff
@@ -141,13 +161,37 @@ internal static TokenCredential CreateCredential(IConfiguration configuration, T
                     throw new InvalidOperationException($"Unable to find a certificate with thumbprint '{certificate}'");
                 }
 
-                var credential = new ClientCertificateCredential(tenantId, clientId, certs[0], identityClientOptions);
+                ClientCertificateCredential credential;
+                if (additionallyAllowedTenantsList != null)
+                {
+                    var options = new ClientCertificateCredentialOptions();
+                    foreach (string tenant in additionallyAllowedTenantsList)
+                    {
+                        options.AdditionallyAllowedTenants.Add(tenant);
+                    }
+                    credential = new ClientCertificateCredential(tenantId, clientId, certs[0], options);
+                }
+                else
+                {
+                    credential = new ClientCertificateCredential(tenantId, clientId, certs[0]);
+                }
+
                 store.Close();
 
                 return credential;
             }
 
             // TODO: More logging
+
+            if (additionallyAllowedTenantsList != null)
```

Comment:
I'm not sure about this part - it relies on the fact that the default CredentialFactory uses DefaultAzureCredential, but this can actually be changed by the user.
https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/extensions/Microsoft.Extensions.Azure/src/Internal/AzureComponentFactoryImpl.cs#L25

Reply:
Ideally, we would flow through any config to the credential factory fallback, but we already do not do this for all of the other settings, which has led to some confusion - https://github.com/Azure/azure-sdk-for-net/issues/30961

Reply:
How was DefaultAzureCredential utilized before?

Reply:
If CreateCredential returned null, AzureComponentFactory will fall back to a user configurable factory. By default the factory will construct a DefaultAzureCredential.
https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/extensions/Microsoft.Extensions.Azure/src/Internal/AzureComponentFactoryImpl.cs#L25

Reply:
Is it valid to provide both a IConfiguration options and a custom `CredentialFactory`?

Reply:
There is nothing preventing someone from doing that. It is exposed through the [UseCredential](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.azure.azureclientfactorybuilder.usecredential?view=azure-dotnet) overloads, which you'd have access to after call [AddAzureClients](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.azure.azureclientservicecollectionextensions.addazureclients?view=azure-dotnet). I don't think it is obvious that this is actually just the fallback and it will be ignored if the supplied configuration matches one of the supported credentials from AzureComponentFactory.

Reply:
It would be nice if we could log when this occurs (along with the other config decisions above), but barring that, perhaps we should just add some additional doc comments to explain some of the flow for these cases. Otherwise, it seems reasonable to default to DefaultAzureCredential here.

Reply:
Yes, we have some work to do as part of resolving https://github.com/Azure/azure-sdk-for-net/issues/30961